pipeline: add prod, devel, and mechanical pipelines

jlebon opened the pull request:
This adds pipeline buildconfigs for all the current streams (testing,
testing-devel, and bodhi-updates). Rather than having a separate
buildconfig for each stream, we just group them by prod, devel, and
mechanical since the triggering within a particular group is likely
going to be the same.

One note here is that I removed the GitHub webhook from the OpenShift
side, and instead made it part of the Jenkinsfile itself. This is
because we need webhooks to Jenkins anyway to trigger off of
`fedora-coreos-config`, so we might as well use the same mechanism for
`fedora-coreos-pipeline` as well.

Another note: I initially had the `prod` pipeline be solely manually
triggered, though changed that later on to be triggered on pushes to
`fedora-coreos-config`: I think it'd be cool if the only manual step
required for a release was simply to push a commit to `testing`. Anyway,
we can refine that later if needed (and all the other triggers as well).

jlebon commented:
OK, dropped WIP on this! It's looking pretty good now I think and works well. I'm sure there's some knobs in there e.g. around triggering that'll need more adjustments. But we can do those as follow-ups as we gain more experience.

jlebon commented:
Requires #78.

dustymabe commented:
I have a nit on the terminology we are using. In this PR we are naming/referring to things as `prod`, `devel` and `mechanical`. We also have a concept in this repo of *developer (could be confused with devel) pipelines*.

I suggest we slightly change our terminology and naming such that it is now `stable`, `devel` and `mechanical`. All of these are classified as `prod`. Anything running in centos CI in the fedora-coreos namespace should be `prod`. We should not allow anything else to be run in that namespace. Anything running anywhere else is a developer pipeline essentially.

So we have the *prod* `stable` streams. *prod* `devel` streams, and *prod* `mechanical` streams. You could individually set up a *devel* `stable` stream, *devel* `devel` stream, etc.

jlebon commented:
Yeah, I've thought about the devel overloading happening too while working on this.

Hmm, how about we just spell it out, i.e. `development`, instead of just `devel` to clear this up? So, `prod`, `development`, and `mechanical`. This actually matches the terminology used in https://github.com/coreos/fedora-coreos-tracker/blob/master/stream-tooling.md as well.

jlebon commented:
> Anything running in centos CI in the fedora-coreos namespace should be prod. We should not allow anything else to be run in that namespace. Anything running anywhere else is a developer pipeline essentially.

Agreed. This is enforced today by explicitly checking for the job name itself in the pipeline. And just plain error out if it doesn't match.

dustymabe commented:
> Yeah, I've thought about the devel overloading happening too while working on this.
> 
> Hmm, how about we just spell it out, i.e. `development`, instead of just `devel` to clear this up? So, `prod`, `development`, and `mechanical`. 

Yeah I think that may help. 

> This actually matches the terminology used in https://github.com/coreos/fedora-coreos-tracker/blob/master/stream-tooling.md as well.

Yeah looking at that document makes me rethink this a bit. Thanks for linking that. 


I think the problem here is we need a clear term to delineate between official fcos streams and a stream that is a side build for a developer.. Maybe we can call official streams official (i.e. change our `prod` var in Jenkinsfiles to `official` , and we could call developer pipelines `hacker` pipelines.

jlebon commented:
> Maybe we can call official streams official (i.e. change our `prod` var in Jenkinsfiles to `official`, and we could call developer pipelines `hacker` pipelines.

Yup sure, WFM.

jlebon commented:
Rebased this on top of #80.

I added a commit which updates the terminology we use in the various places. See the commit message for details. Re-adding WIP on this since I haven't tested it yet.

dustymabe commented on Jenkinsfile.mechanical:
```diff
@@ -0,0 +1,33 @@
+         userRemoteConfigs: [
+            [url: 'https://github.com/coreos/fedora-coreos-config']
+         ],
+         branches: streams.as_branches(streams.prod)
```
streams.production

jlebon replied:
Heh, yeah caught those while testing this morning.

dustymabe commented on Jenkinsfile:
```diff
@@ -100,6 +106,10 @@ podTemplate(cloud: 'openshift', label: 'coreos-assembler', yaml: pod, defaultCon
                 utils.shwrap("""
                 coreos-assembler buildprep s3://${s3_builddir}
                 """)
+            } else if (!prod && utils.path_exists(developer_builddir)) {
```
`!official`

dustymabe commented on Jenkinsfile:
```diff
@@ -176,14 +190,14 @@ podTemplate(cloud: 'openshift', label: 'coreos-assembler', yaml: pod, defaultCon
-              cp -a \$(realpath builds/latest) /srv/devel/${devel_prefix}/build
+              rm -rf ${developer_builddir}
+              mkdir -p ${developer_builddir}
+              cp -aT builds ${developer_builddir}
```
one more `if (prod)` at the bottom of `stage('Archive')`

jlebon commented on devel-up:
```diff
@@ -92,7 +92,7 @@ def get_username():
 
 def process_template(args):
-    params = [f'DEVEL_PREFIX={args.prefix}']
+    params = [f'DEVELOPER_PREFIX={args.prefix}']
```
Done!

dustymabe commented:
after a few fixes I hit this from within my jenkins job:

```
[workspace] Running shell script

+ oc start-build fedora-coreos-pipeline -e STREAM=bodhi-updates

Error from server (Forbidden): buildconfigs.build.openshift.io "fedora-coreos-pipeline" is forbidden: buildconfigs.build.openshift.io "fedora-coreos-pipeline" not found

script returned exit code 1
```

jlebon commented on Jenkinsfile.mechanical:
```diff
@@ -0,0 +1,33 @@
+    pipelineTriggers(streams.get_push_trigger() + [
+        // also run every 30m: we could increase the interval more if we start
+        // triggering on the relevant fedmsgs.
+        cron("H/30 * * * *")
```
Developer pipelines don't have any cron jobs right now. Note the only buildconfig that actually gets created by `devel-up` right now is still just `${developer_prefix}fedora-coreos-pipeline`. And that one was stripped of all triggers.

dustymabe replied:
I think we're glossing over the fact that if I want to run a developer pipeline in a brand new deployment (new project in openshift in any openshift instance) you can't start with `devel-up`. You start with `oc new-app --file=manifests/pipeline.yaml --param=DEVELOPER_PREFIX=$(whoami)-`. In this case you get **everything**, which was OK in the past because we wanted everything (i.e. we used it all for our developer pipelines usecase), but now it's not the same.

jlebon replied:
Yeah, one base issue here is that OpenShift templates don't have conditionals (I'm not sure if Helm or something else would be better here). What we can do is split `pipeline.yaml` into 3:
- `jenkins.yaml` --> S2I + worker base + PVC (and we can split that one out as well when we're ready to make it optional)
- `pipeline.yaml` --> coreos-assembler IS + fedora-coreos-pipeline BC
- `streams.yaml` --> fedora-coreos-pipeline-{production,developer,mechanical} BCs

So then, the workflow would be `oc new-app --file=manifests/jenkins.yaml` and then straight to `./devel-up`, which instantiates `pipeline.yaml`. We could then also have e.g. a `--streams` switch if a developer wants to instantiate streams BCs as well.

jlebon commented:
> after a few fixes I hit this from within my jenkins job:

Right yeah, basically right now, the "wrapper" pipelines don't support developer mode (i.e. it doesn't have a `${developer_prefix}` in its name). To test all this, I just created them all without a prefix in the devel namespace. The main pipeline still failed as expected since it wasn't in the right namespace, but for testing the triggering itself that's fine (which is the only thing that really needs testing in the wrapper pipelines).

I guess we could make those devel aware as well? Though then I think we'd still want `./devel-up` to default to creating just the pipeline.

jlebon commented:
Dropping WIP on this. Tested working well!

dustymabe commented:
> I guess we could make those devel aware as well? 

I think so?

> Though then I think we'd still want `./devel-up` to default to creating just the pipeline.

Some more thoughts in https://github.com/coreos/fedora-coreos-pipeline/pull/76#discussion_r295844134

jlebon commented:
OK, updated this now! `./devel-up` has now evolved into `./deploy`! See commit message for details.

jlebon commented:
>> I guess we could make those devel aware as well?
>
> I think so?

jlebon commented:
>> I guess we could make those devel aware as well?
>
> I think so?

One note about this: I still didn't make the stream pipelines prefixable. Let's do that as a follow-up? I don't think it's something we need the ability to do right away (and of course, one can always do `./deploy --all` in their separate cluster if they want to test them).

dustymabe commented on manifests/pipeline.yaml:
```diff
@@ -52,6 +52,8 @@ objects:
     metadata:
       name: jenkins-s2i
+      annotations:
+        coreos.com/deploy-default: true
```
these annotations are causing me some grief on first startup:

```
Updating:                                                                                
Error from server (BadRequest): error when creating "STDIN": ImageStream in version "v1" cannot be handled as a ImageStream: v1.ImageStream.ObjectMeta: v1.ObjectMeta.Annotations: ReadString: expects " or n, but found t, error found
 in #10 byte of ...|default":true},"labe|..., bigger context ...|ata":{"annotations":{"coreos.com/deploy-default":true},"labels":{"app":"fedora-coreos","template":"f|...
Traceback (most recent call last):           
  File "./deploy", line 235, in <module>      
    sys.exit(main())                         
  File "./deploy", line 38, in main                                                        
    update_resources(args, resources)        
  File "./deploy", line 164, in update_resources                                                                                                                                                                                           check=True, stdout=subprocess.PIPE)      
  File "/usr/lib64/python3.7/subprocess.py", line 487, in run
    output=stdout, stderr=stderr)            
subprocess.CalledProcessError: Command '['oc', 'create', '--filename', '-']' returned non-zero exit status 1.

```

If I delete all the newly added `coreos.com/deploy-default: true` annotations and all `--all` to my `deploy` command then it succeeds.

jlebon replied:
Hmm weird, I didn't hit this on my local server. But yeah, looks like it's complaining about string vs bool.

dustymabe commented:
re: `HACKING.md: simply Jenkins template instantiation`

Thanks for that commit. I didn't know it was a parameter otherwise I would have used it before :)

One nit: `simplify Jenkins template instantiation`

jlebon commented:
Fixes! :arrow_up: 

Tested in a fresh namespace.

dustymabe commented on HACKING.md:
````diff
@@ -180,70 +184,58 @@ reason we create the app first is that otherwise OpenShift will
 ```
 
-This template creates:
+See `./deploy --help` for more information.
````
I kind of liked having all of the parameters explained out here like we were doing before when we listed out all of the parameters of the template. Are you opposed to adding that stanza back but just convert it to reference arguments to deploy rather than template parameters?

jlebon replied:
Done! :arrow_down:

dustymabe commented:
> One note about this: I still didn't make the stream pipelines prefixable. Let's do that as a follow-up?

sounds good

Mostly LGTM - one comment: https://github.com/coreos/fedora-coreos-pipeline/pull/76#discussion_r296285156

dustymabe commented on HACKING.md:
```diff
@@ -180,70 +184,73 @@ reason we create the app first is that otherwise OpenShift will
+- `--pipeline`
+    - Git source URI for pipeline Jenkinsfile.
+- `--config`
+    - Git source URI for FCOS config.
```
since before we had the `_REF` parameters and now we don't  you might want to mention the `@ref` formatting for these two

jlebon replied:
Fixed!

jlebon commented:
Awesome, thanks for all the reviews! Let's get this thing set up for real now! :)
